crates/merkle-tree/src/tree.rs

```diff
@@ -642,7 +645,18 @@ impl<H: FeltHash, const HEIGHT: usize> MerkleTree<H, HEIGHT> {
                     }
                 };
 
-                nodes.push(node);
+                let node_hash = match node_hash_cache.get(&index) {
```

Comment:
I think there's another potential improvement here: since both the serialized node representation and the hash is in the same row in the DB we could just query both at the same time from SQLite essentially for free. The storage abstraction does not make this possible now though, that would require another refactor...

Reply:
Oh, and I don't think we should actually implement this improvement for now. Let's see if the implementation we have is OK (I'd think it is) and improve only if it makes sense.